Xingjian/symco 17

xingjianW opened the pull request:
Done:
1. Create assignment.
2. Assign assignments to individuals or labels.
3. Send an email to notify the assigned students at the specified time.

Todo:
1. The template item cannot be empty, it needs user-friendly prompts.
2. The assigned date cannot be earlier than the time, it needs user-friendly prompts.

Note:
1. In the conversation_templates, my urls conflicted with the urls of the template-manage page, so I renamed my urls to create_assignment_urls.

rosasck committed:
rebase

rosasck committed:
rebase

rosasck committed:
rebase

rosasck committed:
rebase

rosasck committed:
rebase

rosasck committed:
rebase

rosasck committed:
rebase

rosasck committed:
rebase

rosasck committed:
rebase

rosasck committed:
rebase

rosasck committed:
rebase

rosasck committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

JerCarnahan committed:
rebase

kaigoos committed:
rebase

kaigoos committed:
rebase

kaigoos committed:
rebase

kaigoos committed:
rebase

kaigoos committed:
rebase

kaigoos committed:
rebase

kaigoos committed:
rebase

kaigoos committed:
rebase

kaigoos committed:
rebase

kaigoos committed:
rebase

kaigoos committed:
rebase

kaigoos committed:
rebase

devin-w-hill committed:
rebase

devin-w-hill committed:
rebase

devin-w-hill committed:
rebase

devin-w-hill committed:
rebase

devin-w-hill committed:
rebase

devin-w-hill committed:
rebase

devin-w-hill committed:
rebase

devin-w-hill committed:
rebase

devin-w-hill committed:
rebase

devin-w-hill committed:
rebase

devin-w-hill committed:
rebase

devin-w-hill committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
In addition to sending mail other functions have been completed.
Todo: Write a program to send mail.
Advanced: Display the results of fuzzy search through dropbox when entering template information.

xingjianW committed:
In addition to sending mail other functions have been completed.
Todo: Write a program to send mail.
Advanced: Display the results of fuzzy search through dropbox when entering template information.

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
In addition to sending mail other functions have been completed.
Todo: Write a program to send mail.
Advanced: Display the results of fuzzy search through dropbox when entering template information.

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

rosasck commented on simcon_project/conversation_templates/create_assignment_urls.py:
```diff
@@ -0,0 +1,8 @@
+urlpatterns = [
+    path('', CreateAssignmentView.as_view(), name="create_assignment"),
+    path('add_data/', add_data, name="add_data"),
+]
```
Nit: new line

xingjianW replied:
What does "new line" mean?

rosasck commented on simcon_project/conversation_templates/migrations/0003_auto_20210103_0336.py:
```diff
@@ -0,0 +1,112 @@
+# Generated by Django 3.1.3 on 2021-01-03 03:36
```
please delete this from the PR, i think u still dont have all the new main since this is in the git ignore

rosasck commented on simcon_project/conversation_templates/urls.py:
```diff
@@ -10,4 +10,4 @@
     path('templates/delete/<uuid:pk>', TemplateDeleteView.as_view(), name="delete_template"),
     path('folder/remove/<uuid:pk>', remove_template, name="remove_template"),
-]
+]
```
new line

rosasck commented on simcon_project/conversation_templates/views/create_assignment.py:
```diff
@@ -0,0 +1,141 @@
+from django.shortcuts import render
+from users.models import SubjectLabel
```
you can do this on one line so its not 6 lines

rosasck commented on simcon_project/conversation_templates/views/create_assignment.py:
```diff
@@ -0,0 +1,141 @@
+
+#Determine if data is empty.
+def isNull(data):
+    if len(data)!=0:
```
can you chain these with and instead of nested ifs?

rosasck commented on simcon_project/conversation_templates/views/create_assignment.py:
```diff
@@ -0,0 +1,141 @@
+            tempTmp=ConversationTemplate.objects.get(name=tempName,creation_date=tempDate)
+            assignment.conversation_templates.add(tempTmp)
+
+    ##Assign label information to assignment
```
extra #

rosasck commented on simcon_project/conversation_templates/views/create_assignment.py:
```diff
@@ -0,0 +1,141 @@
+    tz = get_localzone()
+    sched = BackgroundScheduler(timezone=tz, job_defaults=job_defaults)
+
+    subject='You received a new assignment.'
```
`Simulated Conversation Assignment Update`

rosasck commented on simcon_project/conversation_templates/views/create_assignment.py:
```diff
@@ -0,0 +1,141 @@
+    subject='You received a new assignment.'
+    msg='You received this email because you have a new assignment: '+name+'. Please check the assignment page.'
+    schedId='Assignment--'+name
+    #run_date='2021-01-22 18:35:00'
```
can you remove this?

xingjianW replied:
This is the format of the date, I think it might be used if we change the sending time of the email in the future.

rosasck commented on simcon_project/conversation_templates/views/create_assignment.py:
```diff
@@ -0,0 +1,141 @@
+
+    # print(success)
+    # print(errMsg)
+    # print(repr(stuIsNull)+"  "+repr(labelIsNull)+"  "+repr(tempIsNull))
```
remove un needed print statements

xingjianW replied:
This is the format of the date, I think it might be used if we change the sending time of the email in the future.

rosasck commented on simcon_project/simcon_project/settings.py:
```diff
@@ -141,8 +142,17 @@
+EMAIL_HOST_USER = 'ccutwang@gmail.com'
+EMAIL_HOST_PASSWORD = 'upxjfdaxtzkmkleq'
+EMAIL_USE_TLS = True
+EMAIL_USE_SSL = False
```
new line

rosasck commented on simcon_project/simcon_project/settings.py:
```diff
@@ -141,8 +142,17 @@
 ]
+
+#EMAIL_BACKEND = 'django.core.mail.backends.console.EmailBackend'
+SENDGID_API_KEY=os.getenv('upxjfdaxtzkmkleq')
```
is this our secret key? or public key? this seems sketchy to not be an .env variable

xingjianW replied:
This is my email password. . . I will delete it.

rosasck commented on simcon_project/simcon_project/settings.py:
```diff
@@ -141,8 +142,17 @@
+
+#EMAIL_BACKEND = 'django.core.mail.backends.console.EmailBackend'
+SENDGID_API_KEY=os.getenv('upxjfdaxtzkmkleq')
+EMAIL_HOST = 'smtp.gmail.com'
```
??

rosasck commented on simcon_project/simcon_project/settings.py:
```diff
@@ -141,8 +142,17 @@
+SENDGID_API_KEY=os.getenv('upxjfdaxtzkmkleq')
+EMAIL_HOST = 'smtp.gmail.com'
+EMAIL_PORT = 587
+EMAIL_HOST_USER = 'ccutwang@gmail.com'
```
whose email is this? and will we get to keep it

rosasck commented on simcon_project/simcon_project/settings.py:
```diff
@@ -141,8 +142,17 @@
+EMAIL_HOST = 'smtp.gmail.com'
+EMAIL_PORT = 587
+EMAIL_HOST_USER = 'ccutwang@gmail.com'
+EMAIL_HOST_PASSWORD = 'upxjfdaxtzkmkleq'
```
this seems very unsafe

rosasck commented on simcon_project/templates/create_assignment.html:
```diff
@@ -0,0 +1,286 @@
+          })
+      }
+  </script>
+</html>
```
would love if you could comment the functions since its a lot of functions or if you could please have a js file!!

rosasck commented on simcon_project/templates/nav_bar.html:
```diff
@@ -0,0 +1,32 @@
+    {% endblock %}
+
+  </body>
+</html>
```
this file no longer exists!

rosasck commented on simcon_project/templates/researcher_view.html:
```diff
@@ -1,7 +1,7 @@
 {% extends "researcher_base.html" %}
 {% load render_table from django_tables2 %}
```
please only git add files you have made code changes so we don't get confused/ break something

rosasck commented on simcon_project/templates/settings_view.html:
```diff
@@ -0,0 +1,35 @@
+<!DOCTYPE html>
```
i deleted this file, it no longer exists, please check your rebase.. seems like u have files that no longer exist.
I think you rebased off of branches and not off main!

rosasck commented on simcon_project/templates/static/researcherTable.css:
```diff
@@ -0,0 +1,28 @@
+.body-container {
```
again, this file is deleted. please slack me and we can fix ur git state. this file was moved

rosasck commented on simcon_project/templates/student_view.html:
```diff
@@ -1,6 +1,6 @@
 {% extends "student_base.html" %}
```
remove from git add

rosasck commented on simcon_project/users/migrations/0005_auto_20210103_0336.py:
```diff
@@ -0,0 +1,102 @@
+# Generated by Django 3.1.3 on 2021-01-03 03:36
```
remove from PR, any migration files should not be pushed up, according to git ignore

rosasck commented on simcon_project/users/views/__init__.py:
```diff
@@ -4,6 +4,7 @@
 from .student_home import StudentView
 from .template_start import TemplateStartView
-from .view_response import ViewResponse, UpdateOverallResponseFeedback, UpdateNodeResponseFeedback
+from .view_response import ViewResponse, UpdateOverallResponseFeedback, UpdateNodeResponseFeedback
```
new line

rosasck commented on simcon_project/users/views/researcher_home.py:
```diff
@@ -1,4 +1,5 @@
 from django.shortcuts import render
+from django.contrib.auth.decorators import login_required
```
you are adding this for a second time, please remove

rosasck commented on simcon_project/users/views/settings_page.py:
```diff
@@ -0,0 +1,7 @@
+from django.shortcuts import render
```
this file no longer exists, please check ur rebases. They were not successful

rosasck reviewed (changes requested):
You have files, that were removes/moved. Please try to pull from main and merge the branches again. Your rebase was not with the correct main state! most of these changes came from my branch. I think you rebased off my branch before i made some heavy changes with naming. Also remove migration files and check your git ignore, that seems to be outdated. and remove any files you did not code into! we all have different machines so of course it will count formatting just don't add them to your PR's!

kaigoos reviewed (changes requested):
I tested out the assignment creation page and there are some things I think need to be changed. 

The assignment creation page should be accessible from the navbar and also be unique to each researcher. The researcher should not have to enter their own email, each students email, each template name, and label name when creating an assignment. That is too troublesome for the researcher to remember all of those things and it is something we can do on the back-end to make this page more user friendly. A multiple choice field like the image below(doesn't have to be exactly like that) should be provided to them for each field to show so the researcher can pick student, templates, and labels more easily.
![image](https://user-images.githubusercontent.com/56237018/106066373-929a5d80-60b1-11eb-94dd-fa7741147054.png)

Also either students or labels need to be optional, both do not need to be required.
Otherwise the page itself looks clean and I like how each thing that is added is shown clearly below that field. The assignments get saved correctly so there is no problem there. There seems to be some validation things missing too, but that can always be done in a separate ticket.

rosasck commented:
Like kai said form the "create assignment view" the request param gives u the current user email! so doing request.email you can figure out which researcher they are

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW commented:
> Like kai said form the "create assignment view" the request param gives u the current user email! so doing request.email you can figure out which researcher they are

I read the content of the request, and it seems that there is no user's email address. Maybe we should store the email address in a cookie?

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

kaigoos commented:
> 
> 
> > Like kai said form the "create assignment view" the request param gives u the current user email! so doing request.email you can figure out which researcher they are
> 
> I read the content of the request, and it seems that there is no user's email address. Maybe we should store the email address in a cookie?

You need to login as a researcher before going to create-assignment page then you can get their email through the request. No need for cookies.

xingjianW commented:
> > > Like kai said form the "create assignment view" the request param gives u the current user email! so doing request.email you can figure out which researcher they are
> > 
> > 
> > I read the content of the request, and it seems that there is no user's email address. Maybe we should store the email address in a cookie?
> 
> You need to login as a researcher before going to create-assignment page then you can get their email through the request. No need for cookies.

Is there any part in your code has this function?

rosasck commented on simcon_project/users/views/researcher_home.py:
```diff
@@ -14,3 +14,4 @@ def ResearcherView(request):
     pageNumber = request.GET.get('page')
     pageObject = paginator.get_page(pageNumber)
     return render(request, 'researcher_view.html', locals())
+
```
remove this file from ur pr

rosasck commented on simcon_project/templates/researcher_view.html:
```diff
@@ -1,6 +1,6 @@
 {% extends "researcher_base.html" %}
```
remove this file from ur PR

rosasck commented on simcon_project/simcon_project/settings.py:
```diff
@@ -143,8 +144,18 @@
     os.path.join(BASE_DIR, 'static/media')
 ]
+
+#EMAIL_BACKEND = 'django.core.mail.backends.console.EmailBackend'
```
is this needed?

xingjianW committed:
Now the program can automatically get the researcher's email address. I checked the implementation of the fuzzy search function on the Internet and found that it is more complicated than I thought. So I plan to do this function next week, but other functions have been perfected.

xingjianW committed:
Now the program can automatically get the researcher's email address. I checked the implementation of the fuzzy search function on the Internet and found that it is more complicated than I thought. So I plan to do this function next week, but other functions have been perfected.

xingjianW committed:
Now the program can automatically get the researcher's email address. I checked the implementation of the fuzzy search function on the Internet and found that it is more complicated than I thought. So I plan to do this function next week, but other functions have been perfected.

xingjianW committed:
Now the program can automatically get the researcher's email address. I checked the implementation of the fuzzy search function on the Internet and found that it is more complicated than I thought. So I plan to do this function next week, but other functions have been perfected.

xingjianW committed:
Now the program can automatically get the researcher's email address. I checked the implementation of the fuzzy search function on the Internet and found that it is more complicated than I thought. So I plan to do this function next week, but other functions have been perfected.

xingjianW committed:
Now the program can automatically get the researcher's email address. I checked the implementation of the fuzzy search function on the Internet and found that it is more complicated than I thought. So I plan to do this function next week, but other functions have been perfected.

xingjianW committed:
Now the program can automatically get the researcher's email address. I checked the implementation of the fuzzy search function on the Internet and found that it is more complicated than I thought. So I plan to do this function next week, but other functions have been perfected.

xingjianW committed:
Now the program can automatically get the researcher's email address. I checked the implementation of the fuzzy search function on the Internet and found that it is more complicated than I thought. So I plan to do this function next week, but other functions have been perfected.

xingjianW committed:
Now the program can automatically get the researcher's email address. I checked the implementation of the fuzzy search function on the Internet and found that it is more complicated than I thought. So I plan to do this function next week, but other functions have been perfected.

xingjianW committed:
Now the program can automatically get the researcher's email address. I checked the implementation of the fuzzy search function on the Internet and found that it is more complicated than I thought. So I plan to do this function next week, but other functions have been perfected.

xingjianW committed:
Now the program can automatically get the researcher's email address. I checked the implementation of the fuzzy search function on the Internet and found that it is more complicated than I thought. So I plan to do this function next week, but other functions have been perfected.

xingjianW committed:
Now the program can automatically get the researcher's email address. I checked the implementation of the fuzzy search function on the Internet and found that it is more complicated than I thought. So I plan to do this function next week, but other functions have been perfected.

xingjianW committed:
Now the program can automatically get the researcher's email address. I checked the implementation of the fuzzy search function on the Internet and found that it is more complicated than I thought. So I plan to do this function next week, but other functions have been perfected.

xingjianW committed:
Now the program can automatically get the researcher's email address. I checked the implementation of the fuzzy search function on the Internet and found that it is more complicated than I thought. So I plan to do this function next week, but other functions have been perfected.

xingjianW committed:
Now the program can automatically get the researcher's email address. I checked the implementation of the fuzzy search function on the Internet and found that it is more complicated than I thought. So I plan to do this function next week, but other functions have been perfected.

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

kaigoos commented on simcon_project/conversation_templates/views/create_assignment.py:
```diff
@@ -0,0 +1,142 @@
+    assignment.date_assigned=date
+    researcher=Researcher.objects.get(email=researcher)
+    assignment.researcher=researcher
+    assignment.save()
```
You are saving the assignment even though you have an error that says "the assigned date cannot be today or before today." Make sure to not save the assignment if it doesn't have a valid date.

xingjianW replied:
I added a judgment at the end of the program. If there is any error during the creation of the assignment, delete the assignment at the end.

kaigoos commented on simcon_project/conversation_templates/urls/create_assignment_urls.py:
```diff
@@ -0,0 +1,8 @@
+app_name = 'assignments'
+urlpatterns = [
+    path('', CreateAssignmentView.as_view(), name="create_assignment"),
+    path('add_data/', add_data, name="add_data"),
```
Make sure to use the naming convention we went over on the weekend. Names should be "create-assignment" and "add-data" with - instead of _

kaigoos reviewed (dismissed):
Some small changes are needed. I feel like creating an assignment and all files related to it should be under "user" and not "conversation_templates", but that's just my preference.

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
Now, if there are any errors in creating a new assignment, the created assignment will be deleted at the end.

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

kaigoos commented on simcon_project/users/views/create_assignment.py:
```diff
@@ -0,0 +1,144 @@
+    data=request.POST
+    name=data.get('name')
+    date=data.get('date')
+    #researcher=request.user
```
You forgot to uncomment the researcher

xingjianW committed:
A small change, added the function of obtaining the researcher's email address that was forgotten before.

xingjianW committed:
A small change, added the function of obtaining the researcher's email address that was forgotten before.

xingjianW committed:
A small change, added the function of obtaining the researcher's email address that was forgotten before.

devin-w-hill commented on simcon_project/templates/create_assignment.html:
```diff
@@ -0,0 +1,268 @@
+          }
+      }
+
+      function add_temp(tbl, name){
```
When I create an assignment, the template I chose doesn't end up getting included along with the assignment.

xingjianW replied:
> When I create an assignment, the template I chose doesn't end up getting included along with the assignment.

Can you describe it dependently? I'm not sure how to improve it.

devin-w-hill replied:
My bad, I must've been working with an old version. Ignore.

devin-w-hill reviewed (dismissed):
erased

devin-w-hill commented on simcon_project/users/views/create_assignment.py:
```diff
@@ -0,0 +1,144 @@
+    dateTmp=datetime.datetime.strptime(date, "%Y-%m-%d")
+    if dateTmp<=today:
+        success=1
+        errMsg=errMsg+'The assigned date cannot be today or before today.\n\n'
```
I read your todo and I'm not sure I read it correctly but I was able to make an assignment with a start date before today's date. So I'm not sure this is working as intended.

devin-w-hill replied:
This works now too.

devin-w-hill commented on simcon_project/users/views/create_assignment.py:
```diff
@@ -0,0 +1,144 @@
+    #when an error occurs, there is no need to add this task to the schedule.
+    if success==0:
+        #run_date='2021-01-22 18:35:00'
+        sched.add_job(sendMail, trigger='date', id=schedId, run_date=date, args=(subject, msg, students, researcher),replace_existing=True)
```
I wasn't able to get an email to come through when I created an assignment

Likhovodov commented on simcon_project/templates/create_assignment.html:
```diff
@@ -0,0 +1,268 @@
+      var addedLabel=new Array();
+
+      function add2table(tblId,name,key,id){
+          console.log(tblId);
```
Delete this and all other console logging

Likhovodov reviewed (dismissed):
I love how clean everything looks, good progress!

My issues:
- Date format should probably be MM--DD--YYYY not YYYY-MM-DD
- You should not be able to add templates, students, and labels that do not belong to that particular researcher.
- If possible, while typing something into the student, label, or template field, all available options should be visible as input is coming in (like a search that shows suggestions). The way it is designed now, exact input is expected which is not ideal)
- Input fields are highlighted in red when they are emptied. This should not happen. When a student/label/template is added, that input field should be cleared and it should not be highlighted red
- I get internal server error when I try to submit to the backend (Error: ConversationTemplate matching query does not exist.)

Likhovodov commented on simcon_project/users/views/create_assignment.py:
```diff
@@ -0,0 +1,144 @@
+import datetime
+
+
+class CreateAssignmentView(TemplateView):
```
Make sure only a researcher can access this view. Maybe look at how is_researhcer(user) from researcher_home.py is used in other views.

xingjianW committed:
Now the researcher can only see the information that belongs to him.

xingjianW committed:
Now the researcher can only see the information that belongs to him.

xingjianW committed:
Now the researcher can only see the information that belongs to him.

xingjianW committed:
Now the researcher can only see the information that belongs to him.

xingjianW committed:
Now the researcher can only see the information that belongs to him.

xingjianW committed:
Now the researcher can only see the information that belongs to him.

xingjianW commented:
> I love how clean everything looks, good progress!
> 
> My issues:
> 
> * Date format should probably be MM--DD--YYYY not YYYY-MM-DD
> * You should not be able to add templates, students, and labels that do not belong to that particular researcher.
> * If possible, while typing something into the student, label, or template field, all available options should be visible as input is coming in (like a search that shows suggestions). The way it is designed now, exact input is expected which is not ideal)
> * Input fields are highlighted in red when they are emptied. This should not happen. When a student/label/template is added, that input field should be cleared and it should not be highlighted red
> * I get internal server error when I try to submit to the backend (Error: ConversationTemplate matching query does not exist.)

xingjianW committed:
The input box will be cleared after clicking the Add button.

xingjianW committed:
The input box will be cleared after clicking the Add button.

xingjianW committed:
The input box will be cleared after clicking the Add button.

xingjianW commented:
> I love how clean everything looks, good progress!
> 
> My issues:
> 
> * Date format should probably be MM--DD--YYYY not YYYY-MM-DD
> * You should not be able to add templates, students, and labels that do not belong to that particular researcher.
> * If possible, while typing something into the student, label, or template field, all available options should be visible as input is coming in (like a search that shows suggestions). The way it is designed now, exact input is expected which is not ideal)
> * Input fields are highlighted in red when they are emptied. This should not happen. When a student/label/template is added, that input field should be cleared and it should not be highlighted red
> * I get internal server error when I try to submit to the backend (Error: ConversationTemplate matching query does not exist.)

1. The reason why the date format is YYYY-MM-DD format is because this is the date input format of the send_mail function.
2. I have added a filter to the query.
3. I am doing it.
4. I have added the function of clearing the input box after clicking the "Add" button.
5. I don't quite understand why this happened. Can you provide more details? In my program, after you click "add template", the program has verified the template information.

xingjianW committed:
Added prompt when typing (dropbox).

xingjianW committed:
Added prompt when typing (dropbox).

xingjianW committed:
Added prompt when typing (dropbox).

xingjianW committed:
Removed the part that transmits sensitive information.

xingjianW committed:
Removed the part that transmits sensitive information.

xingjianW committed:
Removed the part that transmits sensitive information.

xingjianW committed:
rebase, changing the key word of label.

xingjianW committed:
rebase, changing the key word of label.

xingjianW committed:
rebase, changing the key word of label.

xingjianW commented:
All comments have been completed, including prompts when implementing researcher input.
apart from:
1. I don't understand what the last question mentioned by David is. (Error: ConversationTemplate matching query does not exist.)
2. I don't know if I should follow Devin's statement, the assignment date of the assignment can be the date the assignment was created.

devin-w-hill commented:
> I don't know if I should follow Devin's statement, the assignment date of the assignment can be the date the assignment was created.

I did test it out again today and wasn't able to create an assignment with today's date. It seems like a researcher might want to be able to do that if they don't plan ahead.

xingjianW committed:
Researcher can create assignment, if the assignment date is the same day.

xingjianW committed:
Researcher can create assignment, if the assignment date is the same day.

xingjianW committed:
Researcher can create assignment, if the assignment date is the same day.

devin-w-hill commented on simcon_project/users/views/create_assignment.py:
```diff
@@ -0,0 +1,153 @@
+    return listTmp
+
+def sendMail(subject, msg, recipient, researcher):
+    send_mail(subject,msg,researcher,recipient,fail_silently=False)
```
I was having some issues getting emails to send but was able to fix by changing researcher to our email name,                       'simulated.conversation@mail.com'. I think we're switching over to gmail soon but this makes the email sending testable.

devin-w-hill commented on simcon_project/users/views/create_assignment.py:
```diff
@@ -0,0 +1,153 @@
+    msg='You received this email because you have a new assignment: '+name+'. Please check the assignment page.'
+    schedId='Assignment--'+name+'--'+date
+    #when an error occurs, there is no need to add this task to the schedule.
+    if success==0 and not assignToday:
```
The logic here is a little wonky. If success==0 and the assignment date is today, the assignment gets deleted and then an email gets sent to the students telling them about it. I'd check success first, delete accordingly and then schedule a job if the assignment date is not today otherwise send the email now.

devin-w-hill commented on simcon_project/templates/create_assignment.html:
```diff
@@ -0,0 +1,309 @@
+<!DOCTYPE html>
```
I think this should either extend the researcher nav bar or link back to the assignment management page after I've merged it into main.

devin-w-hill reviewed (changes requested):
This is super close to done!

xingjianW committed:
Save the version before the merge. Other functions are normal, but mail cannot be sent.

xingjianW committed:
Save the version before the merge. Other functions are normal, but mail cannot be sent.

xingjianW committed:
Save the version before the merge. Other functions are normal, but mail cannot be sent.

xingjianW committed:
rebase

xingjianW committed:
rebase

rosasck commented on simcon_project/templates/create_assignment.html:
```diff
@@ -0,0 +1,312 @@
+    <link
+      rel="stylesheet"
+      type="text/css"
+      href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/4.4.1/css/bootstrap.min.css"
```
this is no longer needed its in the nav bar

rosasck commented on simcon_project/templates/create_assignment.html:
```diff
@@ -0,0 +1,312 @@
+            <div class="form-group">
+                <label for="add_temp">Add template</label>
+                <div class="add" style="display: flex;">
+                    <input type="text" class="form-control col-lg-7" id="template_name" placeholder="Please input template name." list="temp_list">
```
"Please Input a Template Name"

rosasck commented on simcon_project/templates/create_assignment.html:
```diff
@@ -0,0 +1,312 @@
+            <div class="form-group">
+                <label for="add_stu">Add student</label>
+                <div class="add" style="display: flex;">
+                    <input type="text" class="form-control col-lg-7" id="stu_name" placeholder="Please input student email." list="stu_list">
```
Please Input a Student Emails

ericharper1 replied:
"Please Input a Student Email"

rosasck commented on simcon_project/templates/create_assignment.html:
```diff
@@ -0,0 +1,312 @@
+            <div class="form-group">
+                <label for="add_label">Add label</label>
+                <div class="add" style="display: flex;">
+                    <input type="text" class="form-control col-lg-7" id="label_name" placeholder="Please input label name." list="label_list">
```
Please Input a Label name

rosasck commented on simcon_project/templates/create_assignment.html:
```diff
@@ -0,0 +1,312 @@
+              if(stuTmp.email==stu){
+                  find=true;
+                  if(stuTmp.is_active!=1){
+                      window.alert("Add failed. This user is a not an active user");
```
Adding Studet {insert student name} Failed, They are not a registered User

Likhovodov replied:
*They are not an active user* instead of *They are not a registered User*

ericharper1 replied:
Don't we want a check for is_registered too? Because that field gets flipped only after the student has finished registration.

xingjianW replied:
> Don't we want a check for is_registered too? Because that field gets flipped only after the student has finished registration.

I am not sure about the meaning of "registered". It seems that it only means whether the student has registered. The meaning of "is_active" seems to be whether the student has graduated.

rosasck commented on simcon_project/templates/create_assignment.html:
```diff
@@ -0,0 +1,312 @@
+                  else{
+                      //Determine whether this student is already in the table.
+                      if(addedStu.indexOf(stu)>-1){
+                          window.alert("Adding failed, the student is already in the table.");
```
Adding Student {student name} failed, They have already been added

rosasck replied:
actually should we error? or ignore? why punish researcher for a small mistake

Likhovodov replied:
Ignore is the way to go I think

rosasck commented on simcon_project/templates/create_assignment.html:
```diff
@@ -0,0 +1,312 @@
+                  find=true;
+                  var key=name+"--"+tempTmp.creation_date
+                  if(addedTmp.indexOf(key)>-1){
+                      window.alert("Adding failed, the template is already in the table.");
```
Do you mean assignment?

ericharper1 replied:
No, I think this is the section that takes care of adding the conversation templates.

rosasck commented on simcon_project/templates/create_assignment.html:
```diff
@@ -0,0 +1,312 @@
+              if(labelTmp.label_name==name){
+                  find=true;
+                  if(addedLabel.indexOf(name)>-1){
+                      window.alert("Adding failed, the label is already in the table.");
```
I dont know if we should error, or ignore and keep going

ericharper1 replied:
I think to be consistent with the rest, we should pop up an error.

rosasck commented on simcon_project/templates/create_assignment.html:
```diff
@@ -0,0 +1,312 @@
+
+          //varify name and date
+          if(typeof assign_name == "undefined" || assign_name == null || assign_name == ""){
+              window.alert("Assignment name must not be empty!");
```
Assignments Must Have a Name

rosasck commented on simcon_project/templates/create_assignment.html:
```diff
@@ -0,0 +1,312 @@
+              return;
+          }
+          if(typeof assign_date == "undefined" || assign_date == null || assign_date.split("-").length !=3){
+              window.alert("Assignment start date must in YYYY-MM-DD format and not be empty!");
```
Must have a valid start date. example YYYY-MM-DD

rosasck commented on simcon_project/templates/researcher_view.html:
```diff
@@ -13,7 +13,6 @@
     <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.4.1/css/bootstrap.min.css">
     <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.5.1/jquery.min.js"> </script>
     <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.4.1/js/bootstrap.min.js"></script>
-    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.4.1/js/bootstrap.min.js"></script>
```
please pull in main! this has been changed

rosasck commented on simcon_project/users/create_assignment_urls.py:
```diff
@@ -0,0 +1,8 @@
+app_name = 'assignments'
+urlpatterns = [
+    path('', CreateAssignmentView.as_view(), name="create-assignment"),
+    path('add_data/', add_data, name="add-data"),
```
what is add data, i suggest "add_assignment". data is too broad

rosasck commented on simcon_project/users/views/create_assignment.py:
```diff
@@ -0,0 +1,156 @@
+from django.core import serializers
+from django.http import HttpResponse
+from django.core.mail import send_mail
+from apscheduler.schedulers.background import BackgroundScheduler
```
i looked into this 
https://docs.djangoproject.com/en/2.2/howto/custom-management-commands/#howto-custom-management-commands
https://stackoverflow.com/questions/573618/set-up-a-scheduled-job
and this seems the way to do it!

rosasck commented on simcon_project/users/views/create_assignment.py:
```diff
@@ -0,0 +1,156 @@
+    if stuIsNull:
+        if labelIsNull:
+            success=1
+            errMsg=errMsg+'One of the students or labels must not be empty.\n\n'
```
You must have at least one student or label added

xingjianW replied:
The main purpose of this step is to determine whether the student data should be written into the assignment. The content of the error message is that both cannot be empty, so the order of judgment does not affect the result. If the student is empty and the label is not empty, the data of the label will be added to the assignment when it runs to the part responsible for adding the label.

rosasck commented on simcon_project/users/views/create_assignment.py:
```diff
@@ -0,0 +1,156 @@
+    #Assign template information to assignment
+    if tempIsNull:
+        success=1
+        errMsg=errMsg+'Template must not be empty.\n\n'
```
Must include at-least one template

rosasck commented on simcon_project/users/views/create_assignment.py:
```diff
@@ -0,0 +1,156 @@
+    tz = get_localzone()
+    sched = BackgroundScheduler(timezone=tz, job_defaults=job_defaults)
+
+    subject='Simulated Conversation Assignment Update'
```
`New Simulated Conversation Assignment`

rosasck commented on simcon_project/users/views/create_assignment.py:
```diff
@@ -0,0 +1,156 @@
+    msg='You received this email because you have a new assignment: '+name+'. Please check the assignment page.'
+    schedId='Assignment--'+name+'--'+date
+    #when an error occurs, there is no need to add this task to the schedule.
+    researcher='simulated.conversation@mail.com'
```
change to davids email he has in the settings.py

Likhovodov replied:
It's in main so should be correct once he brings it up to speed with main

rosasck commented on simcon_project/users/views/create_assignment.py:
```diff
@@ -0,0 +1,156 @@
+            sendMail(subject, msg, students, researcher)
+        else:
+            #To test the sending mail function, change the content of run_date to the time you want.
+            #run_date='2021-01-22 18:35:00'
```
Do you get to choose the time?

xingjianW replied:
> Do you get to choose the time?

I am not quite sure what you mean. We can change the time of sending the email by changing the run_date in the code. This is a function that may be used during testing. Because the specific time of assignment cannot be set from the researcher's perspective, only the assigned date can be selected. By default, the mail will be sent at midnight, which is not convenient for testing. Changing the run_date in the code can set the time of sending the email to two minutes later.

rosasck commented on simcon_project/users/views/researcher_home.py:
```diff
@@ -18,3 +18,4 @@ def researcher_view(request):
     page_number = request.GET.get('page')
```
this has been changed, pull main in

rosasck commented on simcon_project/templates/create_assignment.html:
```diff
@@ -0,0 +1,312 @@
+<!DOCTYPE html>
+{% extends "researcher_base.html" %}
+<html lang="en">
```
also you need content blocks or it will render your stuff wierd surround with render content blocks

ericharper1 commented on simcon_project/simcon_project/urls.py:
```diff
@@ -47,6 +47,7 @@
     path('researcher/response/', view_response, name="view-response"),
     path('researcher/response/update/<uuid:pk>/', update_overall_response_feedback, name='update-overall-response-feedback'),
     path('researcher/response/updatenode/<uuid:pk>/', update_node_response_feedback, name='update-node-response-feedback'),
+    path('researcher/create-assignment/',include('users.create_assignment_urls'), name="create-assignment-view"),
```
Need a single whitespace between ',' and 'include' :)

Commit:
Minor updates

Commit:
rebase

Commit:
Merged main

Commit:
Updated requirements.txt

Commit:
Updated requirements.txt

Commit:
Added code for datetimepicker on create assignment page.

Commit:
Added code for datetimepicker on create assignment page.

devin-w-hill committed:
Eric added calendar. Working on hooking it up with date and time.

devin-w-hill committed:
rebase

xingjianW committed:
Made some fixes to the code based on the comments of the PR.

xingjianW committed:
Made some fixes to the code based on the comments of the PR.

xingjianW committed:
Made some fixes to the code based on the comments of the PR.

xingjianW committed:
Made some fixes to the code based on the comments of the PR.

xingjianW committed:
Merge remote-tracking branch 'origin/Xingjian/SYMCO-17' into Xingjian/SYMCO-17

# Conflicts:
#	simcon_project/templates/create_assignment.html
#	simcon_project/users/create_assignment_urls.py
#	simcon_project/users/views/create_assignment.py

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
rebase

xingjianW committed:
Merge remote-tracking branch 'origin/Xingjian/SYMCO-17' into Xingjian/SYMCO-17

# Conflicts:
#	simcon_project/templates/create_assignment.html
#	simcon_project/templates/researcher_view.html

xingjianW committed:
Improved the problem of adding label and timing function.

devin-w-hill committed:
calendar changes to html, added django-apscheduler to requirements. moved urls for create assignment and researcher settings. new app for scheduler

devin-w-hill committed:
Changed back to calendar widget. Client side verification to make sure assign now or calendar date picked. If assigning now, email sends. Otherwise job created. TODO: implement scheduler

devin-w-hill committed:
added Email model for scheduler to check against and create an Email alongside any assignments that aren't assigned immediately.

kaigoos reviewed (commented):
There are some minor changes needed. Unregistered students show up in the list of suggestions, the suggestions for labels shows data not related to labels (suggests words that I have typed before in other places), and also the "assign now" option still prompts an alert and says I can't make an assignment.

xingjianW committed:
Solved the problem in the PR comment.
1. Filter the suggestion list
2. You can use "assign_now"
3. The recipient of the mail in the case of "assign_now" now contains the students in the label.

devin-w-hill committed:
fixed assign now not sending emails

devin-w-hill committed:
updated email model to delete along with their assignment.

Commit:
Added input fields for toggle settings and persistence in backend.

xingjianW committed:
Merge branch 'main' into Xingjian/SYMCO-17

Commit:
Merge branch 'main' into Xingjian/SYMCO-17

kaigoos committed:
Fixed date display on templates table in assignment creation

devin-w-hill committed:
date input field toggles with assign now checkbox.

devin-w-hill committed:
Merge remote-tracking branch 'origin/Xingjian/SYMCO-17' into Xingjian/SYMCO-17

Commit:
Adding response attempts field for create assignment page.

Commit:
Merge branch 'Xingjian/SYMCO-17' of github.com:Likhovodov/Simulated-Conversations into Xingjian/SYMCO-17

kaigoos committed:
autocomplete turned off

kaigoos committed:
Merge branch 'Xingjian/SYMCO-17' of https://github.com/Likhovodov/Simulated-Conversations into Xingjian/SYMCO-17

Commit:
Reworked validation for attempts fields.

Commit:
Added dimming of display and other minor changes

Commit:
Brought up to date with origin

Commit:
Organize.

Commit:
Merge branch 'Xingjian/SYMCO-17' of github.com:Likhovodov/Simulated-Conversations into Xingjian/SYMCO-17

Commit:
Updated crate assignment page to use selects

Commit:
Merged remote in

Commit:
Moved django cleanup app import

Commit:
Merge branch 'Xingjian/SYMCO-17' of github.com:Likhovodov/Simulated-Conversations into Xingjian/SYMCO-17

ericharper1 reviewed (approved):
This is ready to go!
